\documentclass[10pt]{article}
\usepackage{longtable}
\usepackage{ulem}
\usepackage{authblk}
\usepackage{algorithmic}
\def\includegraphic{}
\def\includegraphics{}
\pagestyle{myheadings}

\setlength{\parindent}{0pt}
\setlength{\parskip}{1ex plus 0.5ex minus 0.2ex}

\usepackage[paper=letterpaper,margin=1in]{geometry}

\title{phRAIDER: Pattern-Hunter based Rapid Ad Initio \\ Identification of Elemenrary Repeats}
\author[1]{Carly E. Schaeffer}
\author[1]{Nathniel D. Figueroa}
\author[2]{Xiaolin Liu}
\author[1,2,3,4]{\\John E. Karro\thanks{Corresponding Author}}
\affil[1]{Department of Computer Science and Software Engineering}
\affil[2]{Cell, Molecular, and Structural Bology}
\affil[3]{Department of Microbiology}
\affil[4]{Department of Statisticcs, Miami University, Oxford, Ohio (USA)}
\renewcommand\Affilfont{\itshape\small}

\date{}

\begin{document}
\maketitle
\abstract{This is the abstract.}

\section*{Introduction}

Transposable Elements (TEs) are genomic sequences thathad the capacity
to insert copies of themselves into other genomic locations, resulting
in homologous families of sequences spread across the genome.  Present
in almost every higher order genome (covering as much as 45\% of the
human genome and 90\% of the maize genome
\cite{Venter:2001p92,SanMiguel:1996wa}), TEs have proved an important
source of data in numerous studies of genomic structure (e.g.
\cite{Arndt:2005p279,Karro:2008p123,Mugal:2009p581,Hardison:2003p97}).
But given their prevalence, it is important for those studying other
aspect the genome to have TEs masked out -- their bases replaced by N
to allow for easy identification and filtering.  Failure to filter can
reek havoc with genomic analysis tools.  For example, unfiltered TEs can trigger
huge numbers of false positives in automated gene finding tool
\cite{Jiang:2013jt}, as well as inflate tool runtime.

The best tools for repeat identification are RepeatMasker and nHMMER
\cite{RepeatMaskerOpen:XkNxXSd7,Wheeler:2013gj}, but both employ 
library-based search strategies using a pre-compiled
description of sequences in the family (e.g. a ancestral sequence for
BLASTing, or a profile HMM).  But, much like we ask how
the snow plow driver gets to work \cite{Pratchett:uw}, we must ask how
these libraries are compiled.  Library-based tools
are useless for the discovery of new families, and this are
challanging to use on newly sequenced genomes.

\begin{algorithmic}
\If {$i\geq maxval$}
    \State $i\gets 0$
\Else
    \If {$i+k\leq maxval$}
        \State $i\gets i+k$
    \EndIf
\EndIf
\end{algorithmic}


\bibliographystyle{bmc-mathphys} % Style BST file (bmc-mathphys, vancouver, spbasic).
\bibliography{bmc}      % Bibliography file (usually '*.bib' )



\end{document}
