%% BioMed_Central_Tex_Template_v1.06
%%                                      %
%  bmc_article.tex            ver: 1.06 %
%                                       %

%%IMPORTANT: do not delete the first line of this template
%%It must be present to enable the BMC Submission system to
%%recognise this template!!

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                     %%
%%  LaTeX template for BioMed Central  %%
%%     journal article submissions     %%
%%                                     %%
%%          <8 June 2012>              %%
%%                                     %%
%%                                     %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                                                 %%
%% For instructions on how to fill out this Tex template           %%
%% document please refer to Readme.html and the instructions for   %%
%% authors page on the biomed central website                      %%
%% http://www.biomedcentral.com/info/authors/                      %%
%%                                                                 %%
%% Please do not use \input{...} to include other tex files.       %%
%% Submit your LaTeX manuscript as one .tex document.              %%
%%                                                                 %%
%% All additional figures and files should be attached             %%
%% separately and not embedded in the \TeX\ document itself.       %%
%%                                                                 %%
%% BioMed Central currently use the MikTex distribution of         %%
%% TeX for Windows) of TeX and LaTeX.  This is available from      %%
%% http://www.miktex.org                                           %%
%%                                                                 %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%% additional documentclass options:
%  [doublespacing]
%  [linenumbers]   - put the line numbers on margins

%%% loading packages, author definitions

%\documentclass[twocolumn]{bmcart}% uncomment this for twocolumn layout and comment line below
\documentclass{bmcart}

%%% Load packages
%\usepackage{amsthm,amsmath}
%\RequirePackage{natbib}
%\RequirePackage[authoryear]{natbib}% uncomment this for author-year bibliography
%\RequirePackage{hyperref}
\usepackage[utf8]{inputenc} %unicode support
%\usepackage[applemac]{inputenc} %applemac support if unicode package fails
%\usepackage[latin1]{inputenc} %UNIX support if unicode package fails


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                             %%
%%  If you wish to display your graphics for   %%
%%  your own use using includegraphic or       %%
%%  includegraphics, then comment out the      %%
%%  following two lines of code.               %%
%%  NB: These line *must* be included when     %%
%%  submitting to BMC.                         %%
%%  All figure files must be submitted as      %%
%%  separate graphics through the BMC          %%
%%  submission process, not included in the    %%
%%  submitted article.                         %%
%%                                             %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


\def\includegraphic{}
\def\includegraphics{}



%%% Put your definitions there:
\startlocaldefs
\endlocaldefs


%%% Begin ...
\begin{document}

%%% Start of article front matter
\begin{frontmatter}

\begin{fmbox}
\dochead{Research}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% Enter the title of your article here     %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\title{Fast de novo transposable element annotation}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% Enter the authors here                   %%
%%                                          %%
%% Specify information, if available,       %%
%% in the form:                             %%
%%   <key>={<id1>,<id2>}                    %%
%%   <key>=                                 %%
%% Comment or delete the keys which are     %%
%% not used. Repeat \author command as much %%
%% as required.                             %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\author[
   addressref={aff1},
   noteref={n1},
   email={schaefce@miamiOH.edu}
]{\inits{SE}\fnm{Carly E} \snm{Schaeffer}}
\author[
   addressref={aff1},                   % id's of addresses, e.g. {aff1,aff2}
   noteref={n1},                        % id's of article notes, if any
   email={figuernd@miamiOH.edu}   % email address
]{\inits{ND}\fnm{Nathaniel D} \snm{Figueroa}}
\author[
   addressref={aff1,aff2},
   email={liux17@miamiOH.edu}
]{\inits{X}\fnm{Xiaolin} \snm{Liu}}
\author[
  addressref={aff1,aff2,aff3,aff4},
  corref={aff1},
  noteref={n2},
  email={karroje@miamiOH.edu}
]{\inits{JE}\fnm{John E} \snm{Karro}}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% Enter the authors' addresses here        %%
%%                                          %%
%% Repeat \address commands as much as      %%
%% required.                                %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\address[id=aff1]{%                           % unique id
  \orgname{Department of Computer Science and Software Engineering}, % university, etc
  %\street{Miami University},                     %
  %\postcode{}                                % post or zip code
  %\city{Oxford},                              % city
  %\cny{USA}                                    % country
}
\address[id=aff2]{%                           % unique id
  \orgname{Center for  Molecular and Structural Biology},
  %\street{Miami University},                     %
  %\postcode{}                                % post or zip code
  %\city{Oxford},                              % city
  %\cny{USA}                                    % country
}
\address[id=aff3]{%                           % unique id
  \orgname{Department of Statistics}, % university, etc
  %\street{Miami University},                     %
  %\postcode{}                                % post or zip code
  %\city{Oxford},                              % city
  %\cny{USA}                                    % country
}
\address[id=aff4]{%                           % unique id
  \orgname{Department of Microbiology}, % university, etc
  \street{Miami University},                     %
  %\postcode{}                                % post or zip code
  \city{Oxford},                              % city
  \cny{USA}                                    % country
}


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% Enter short notes here                   %%
%%                                          %%
%% Short notes will be after addresses      %%
%% on first page.                           %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\begin{artnotes}
%\note{Sample of title note}     % note to the article
  \note[id=n1]{Equal contributor} % note, connected to author
  \note[id=n2]{Corresponding Author}
\end{artnotes}

\end{fmbox}% comment this for two column layout

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% The Abstract begins here                 %%
%%                                          %%
%% Please refer to the Instructions for     %%
%% authors on http://www.biomedcentral.com  %%
%% and include the section headings         %%
%% accordingly for your article type.       %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\begin{abstractbox}

\begin{abstract} % abstract
\parttitle{Background} The problem of {\it de novo} identification of
transposable elements (the discovery and annotation of transposable
elements without the use of a pre-complied profile) has been addressed
by a number of tools -- all of which are either based on
computationally complex algorithms that cannot scale to whole-genome
use, or whose sensitivity suffers significantly from the presence of
sequence variation.  Here we present phRAIDER (Pattern Hunter based
Rapid Ab Initio Detection of Elementary Repeats) for quickly masking a
newly sequence, or poorly annotated, genome of transposable elements
that is robust to the sequence variation present in real data.

\parttitle{Results}

\parttitle{Conclusions}

\end{abstract}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% The keywords begin here                  %%
%%                                          %%
%% Put each keyword in separate \kwd{}.     %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\begin{keyword}
\kwd{transposable elements}
\kwd{elementary repeats}
\kwd{pattern hunter}
\kwd{genomic masking}
\end{keyword}

% MSC classifications codes, if any
%\begin{keyword}[class=AMS]
%\kwd[Primary ]{}
%\kwd{}
%\kwd[; secondary ]{}
%\end{keyword}

\end{abstractbox}
%
%\end{fmbox}% uncomment this for twcolumn layout

\end{frontmatter}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% The Main Body begins here                %%
%%                                          %%
%% Please refer to the instructions for     %%
%% authors on:                              %%
%% http://www.biomedcentral.com/info/authors%%
%% and include the section headings         %%
%% accordingly for your article type.       %%
%%                                          %%
%% See the Results and Discussion section   %%
%% for details on how to create sub-sections%%
%%                                          %%
%% use \cite{...} to cite references        %%
%%  \cite{koon} and                         %%
%%  \cite{oreg,khar,zvai,xjon,schn,pond}    %%
%%  \nocite{smith,marg,hunn,advi,koha,mouse}%%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\newtheorem{definition}{Definition}
\newtheorem{observation}{Observation}
\newtheorem{lemma}{Lemma}
\newtheorem{theorem}{Theorem}

%%%%%%%%%%%%%%%%%%%%%%%%% start of article main body
\section*{Background}
Transposable elements, or TEs, are genomic sequences that have at some
point had the capacity to insert copies of themselves into other
genomic locations, resulting in homologous families of sequences
spread across the genome.  TEs are present in almost every higher
order genome, covering as much as 45\% of the human genome and 90\% of
the maize genome [CHECK THIS] \cite{Venter:2001p92} [CITE MAIZE].  TEs
have proved an importent sorce of data in a number of studies
\cite{Arndt:2005p279,Karro:2008p123,Mugal:2009p581,Hardison:2003p97},
but given their prevalence, it is just important for those studying
the genome in them to have them masked out -- their bases replaced by
N to allow for easy identification and filtering.  For example. if not
filtered, TEs can trigger huge numbers of false positives automated
gene finding tool \cite{Jiang:2013jt}, as well as inflate tool
runtime. 

The most well known tools for repeat identification are RepeatMasker
and nHMMER \cite{RepeatMaskerOpen:XkNxXSd7,Wheeler:2013gj}, both of
which work on the principle of a library-based search.  That is, to
identify transposable elements to a known family, the tools require
some description of sequences in the family (e.g. a ancestral sequence
for BLASTing, or a profile HMM description) that is used as the basis
for identification.  But, much like we ask how the snow plow driver
gets to work, or how dictionary authors look up the spelling of a word
\cite{Pratchett:uw}, we must ask how the libraries on which the tools are based are
initially compiled.  Such tools are useful for the discovery of new
families, and cannot be applied to a newly sequenced genome for which
such a library has not been compiled.


Within mammalian species we can, to some extent, rely on homology
relationships to port libraries across species.  A library description
of a human ALU sequence can likely be used for the initial annotation
of a newly sequenced primate genome, and many human LINEs will extend to
rodent, or even more distant, mammalian species.  This does not hold
so well in plants: in many cases TE composition of a given plant organism
is species-specific.  For example, a rice-based TE library will only identify 25\%
of the TEs in the maize genome \cite{Jiang:2013jt}.

To solve this problem we turn to {\it de novo} TE identification
tools: tools identifying TEs though a genome using only the genome
sequence information.  A number of such tools are discussed in the
literature, operating on a number of different principles.  Tools such
as RECON and PILER are based on
self-alignment, using WU-BLAST and LASTZ for the alignment
\cite{Bao:2002,Edgar:2005p2365,Lopez:2003td,Harris:2007uf}.  RECON show good sensitivity but is
computationally intensive and infeasible for use on whole genomes
(requiring 60 hours for 18Mb rice genome int 2013 \cite{Jiang:2013jt}), while
PILER achieves a good runtime with very low sensitivity
\cite{Jiang:2013jt}.  ReAS \cite{Li:2005p3138} and RepeatScout
\cite{Price:2005p1247} are based on $k$-mer searches, with the
earlier showing less sensitivity than RECON \cite{Jiang:2013jt}.
Finally RepeatGluer \cite{Pevzner:2004p3157,Zhi:2006p3199} is
based on a variation of DeBrujin graphs, which allows for a
decomposition of TE families into domains, but is very, very
computationally expensive.  In Saha {\it et al.} the authors perform
an extensive comparison of the tools, and conclude that RepeatScout is
the best tool overall for assembled genomes, while ReAS the best when
dealing with unassembled sequence fragments \cite{Saha:2008dm}.

[TO DO: Make sure we mention somewhere that RAIDER should work for
  unassembled genomes.]

\subsection*{Elementary Repeats}
Another line of development, first proposed by Zheng and Lonardi
\cite{Zheng:2005bl}, involves the use of {\it elementary repeats}.
Similar to the RepeatGluer domains, elementary repeats are
decompositions of TEs into basic building blocks.  Identification of
these building blocks are sufficient for the purpose
of masking, and can be assembled into Transposible Elements for those
interested in studying those elements.

While it is noteriously difficult to state a mathematical definition
of elementary repeats \cite{Bao:2002}, elemenary repeats are a bit
easier to model.  For a given genome, a nucleotide sequence $r$ is an
elementary repeat if:
\begin{enumerate}
\item It is of at least length $l$.  (The length requirement.)
\item At least $f$ exact copies of $r$ appear in the genome.  (The
  frequency requirement.)
\item There is no proper substring of $r$ that is of at least length
  $l$ and appears in the genome independently of $r$. (The minimality
  requirement.)
\item There is no string that contains $r$, appears in the genome
  the same number of times as $r$, and satisfies conditions 1-3.  (The
  maximality requirement.)
\end{enumerate}
The first two criteria are straight-foward, allowing us to designate a
threshold length and frequency for our model.  The minimality
requirement ensures that an elementary repeat does not contain
anything that could be viewed as an independent elementary repeat --
that it is the ``smallest building block'' that we can find.  While
the maximality requirement ensures that it is not tied to some other, larger,
elementary repeat.


Having proposed this definition, Zheng and Lonardi developed an
identification algorithm that had a runtime quadratic in the genome
size -- so not of practical use on whole-genome analysis.  This was
refined to linear time by He and also by Huo {\it et al.}
\cite{He:2006gpa,Huo:2009hoa} based on variations of suffix tree
approaches, but because of that use of suffix trees they are limited
in their ability to handle sequence variation.  As transposable
elements naturally suffer from copy mistakes and accumulate
instance-specific base substitutions over time, this is a significant
limitation.

\subsection*{RAIDER}
\label{RAIDERSec}
It was with the objective of creating a linear time identification
algorithm that could handle variation through use of
PatternHunter-like spaced seeds that we developed the prototype RIADER
\cite{Li:2004wl}.  A rough implementation was first presented in
Figueroa {\it et al.}, with more details in the Figueroa masters
thesis \cite{Figueroa:2014uk,Figueroa:2013cz}.  Not itself fully able
to handle the spaced seed component, RAIDER was built along an
alternate, but equivalent definition of elementary repeats based on
$l$-mers (sequences of length $l$).  Specifically, it was observed
that the minimality condition could be rewritten as: 
There is no $l$-mer contained with $r$ that appears in the genome
more times than $r$.  From there we make our four core observations that form
the basis for the RAIDER algorithm:
\begin{enumerate}
\item An $l$-mer cannot belong to two different elementary repeats.
\item Any $l$-mer in the genome that occurs $f$ or more times is
  either an elementary repeat or belongs to one.
\item Any two $l$-ers belonging to an elementary repeat must appear
  the same number of times in the genome. 
\item Given two sequences in the genome that are {\it maximally
  identical} (that is, cannot be extended in either direction and still
  be identical), these sequences cannot belong to a larger
  elementary repeat.
\end{enumerate}
\end{observation}
By ``belong'' we mean ``is a substring of''; the reason for the
terminology will be explained shortly.  For discussion and proof, see
the Figueroa Thesis \cite{Figueroa:2013cz}. 

Based on these observations, we discover we can find all elementary
repeats in a single scan of the genome.  Specifically, as we scan from
left to right, we track $l$-mer occurrences and identify multiple
copies of the same $l$-mer.  When we find the same sequence of $l$-mers
occurring multiple times in a row, we can mark it as a tentative family,
then break it down later if we discover violations of the minimality
condition.  The algorithm is summarized in Figueroa {\it et al.}, and
discussed in detail in the Figueroa Thesis \cite{Figueroa:2014uk,Figueroa:2013cz}.

Results of the preliminary implementation were promising.  Based on
the Saha {\it et al} analysis putting RepeatMasker as the top {\it de
  novo} search tool, comparisons were against that.  RAIDER result
quality was close, but slightly behind -- with runtime
order of magnitudes better.  On human chromosome 22 we saw a
$12\times$ speedup from RepeatScout to RAIDER (2344 seconds to 192 seconds), while coverage
of the RepBase ancesteral sequecne \cite{Jurka:2005bl} actually improved (77\% to 84\%), while on mouse chromosome 19 we saw 
the same speedup with a significant drop in coverage (53\% to 30\%).  On
the full human genome RAIDER ran in $6.3$ hours, while RepeatScout was
unable to complete it run.  For details, see Figueroa {\it et al.} \cite{Figueroa:2014uk}.

\subsection*{Spaced Seeds}
 
PatternHunter, first proposed as a very successful augmentation to
BLAST \cite{Li:2004wl,Altschul:1997p843}, is based on the notion of {\it
  spaced seeds}.  In the context of BLAST, instead of looking for a
contiguous substring match to trigger the alignment extension,
PatternHunter uses a more refined match pattern.  For example, the
seed 1111110111111 would specify that we were looking for an exact
match of two length-six substrings separated by a single base (which
might or might not match).  In the toy seed $s=110101$, we would say the
strings $AACACA$ and $AAGACA$ because all positions corresponding to a
1 match.  The string $AACACA$ does not math $TACACA$ (with respect to
$s$), as the single miss-match does not correspond to the 1.  Nor does
$AACACA$ match any substring of $TTAACTCA$, though it does match a
substring of $T{\mathbf AAGATA}C$.

The incorporation of spaced seeds into BLAST lead to significant
improvements at virtually no cost in runtime \cite{Li:2004wl}.
SpecificallyIn our context, we will use them to allow limited
miss-matches between instances of ``the same'' elementary repeat.  The
preliminary version of RAIDER from Figueroa {\it et al.} made some
attempt to incorporate this -- was successful enough to provide a
proof-of-concept, but a significant amount of work was left to be
done.

In this paper we present phRAIDER (PatternHunter-based RAIDER),
implemented as open source code that may be obtained [NEED WEB
  ADDRESS].  In our Methods section we present the new model that
allows us to extend RAIDER to correctly use spaced seeds, and briefly
outline the algorithm (with more details provided in the supplmentary
materials), with a analsis of phRAIDER performance in our Results.


\section*{Methods}

In order to allow the use of a spaced seed strategy, we will need to
define a new model of elementary repeats that will accommodate it.  In
the following we will first outline our theoretical framework
(described in detail in the Supplementary Materials), and then review
the phRAIDER algorithm.

\subsection*{Seeded Elementary Repeats}
In order to present our new model of elementary repeats, we first need
to define a few terms:
\begin{enumerate}
  \item A spaced seed is a binary string that starts and ends with 1
    symbols (as defined in Li {\it et al.} \cite{Li:2004wl}).
  \item A {\it sequence descriptor} is a DNA string that allows *
    symbols (indicating that a position where base content does not
    matter).  A string matches a sequence decriptor of equal length if
    it has the same base in all non-* positions.
  \item We say the frequency of $r$ in a genomic sequence $G$ (denoted
    $\nu_G(r)$) is the number of subsequences in $G$ that match $r$.
  \item A spaced seed $s$ {\it hits} a substring of sequence
    descriptor $r$ if that substring is the same length of $s$ and,
    when aligned, every $*$ in the substring matches a 0 in the
    string.  (Example: if $s=11011$ and $r=AAA*TTTTT$, then $s$
    hits both $AA*CC$ and $TTTTT$, but not $A*TTT$.)
  \item The {\it decomposition} of $r$ (with respect to $s$) is the set of all substrings of
    $r$ that are hit by $s$.  For the {\it generalized decomposition},
    we take each member of the decomposition and replace by $*$ any
    base that aligned to a 0 in the seed.  (Example: if $s=11011$ and
    $r=AAATTT$, then $AAATT$ is in the decomposition, so $AA*TT$ is in
    the generalized decomposition.)
  \item We say a string $s$ {\it covers} a sequence descriptor $r$ if
  every base in $r$ is contained within at least one substring from
  the decomposition of $r$.
\end{enumerate}

The Zheng and Lonardi elementary repeats are all identical, hence can
be described by a single string.  Because we allow variation, we
required the flexibility of the sequences descriptor.  Leading us to
our definition:

\begin{definition}
  \label{REDDef}
  Given a fixed genomic sequence $G$, a integer $f$, and spaced seed $s$, a sequence
  descriptor $r$ is a {\it repeat element descriptor} if:
  \begin{enumerate}
  \item $s$ covers $r$.  (Minimum length requirement.)
  \item $\nu_G(r) \geq f$.  (Frequency requirement.)
  \item For every string $w$ in the generalized decomposition of $r$
    (w.r.t. $s$), $\nu_G(w) = \nu_G(r)$.  (Minimality requirement.)
  \item There is no sequence descriptor $r'$ such that (1) $r$ is
    a substring of $r'$, (2) $\nu_G(r') = \nu_G(r)$, and (3) $r'$
    satisfies conditions 1-3.  (Maximality requirement.)
  \end{enumerate}
\end{definition}

We will refer to the set of sequences in the genome that match the
repeat family descriptor to be the repeat element family of the
descriptor.

\begin{theorem}
Definition~\ref{REDDef} is equivalent to the Z\&L
definition of elementary repeats when $s$ contains no 0 symbols.
\end{theorem}
In the Supplementary materials we prove our definition equivalent to
the Figueroa definition (Definition 5 from the thesis) -- which is
proved to be equivalent to the Z\&L definition in that work \cite{Figueroa:2014uk}.

It is instructive to look at a specific example.  Consider the toy
seed $11011$, let $f=2$, and assume the genome contains a unique
instance of $g_1 = AAAAA{\mathbf C}TTTTT$ and a unique instance of
$g_2 = AAAAA{\mathbf G}TTTTT$.  Consider the sequence descriptor $r_1
= AAAAA*TTTTT$.  A simple inspection shows us that conditions (1)-(3)
of Definition~\ref{REDDef} hold, so if we assume the surrounding
sequence is such that (4) holds as well, this is an elementary repeat
descriptor for this family of sequences.  (Note $AA*AA*TT*TT$ -- the
elementary repeat descriptor is not unique.)  But if we add a a second
instance of $g_3 = AAAAAGTTTTT$, something happens that was not
possible in the unseeded version.  While $r_1$ continues to meet the definition, the $g_2/g_3$
subsequence $AAAAGTTTT$ also induces an elementary repeat family
$r_2 = AAAA*TTTT$. (Note that we cannot extend this by, for example,
one base to the left, as the $AAAAA$ substring would cause a violation
of condition (3)).  So $r_2$ corresponds to strings in $G$ that are
substrings of those that $r_1$ matches -- a condition not possible
with the seedless elementary repeats.  This also then violates one of
our core observation on which RAIDER is broken: we now have an $l$-mer
($AAAA*$) that belongs to multiple descriptors.

A second complicating factor is that our repeat element descriptor is
not a sequence of consecutive $l$-mers. In the seedless version, an
elementary repeat of length $n$ can be viewed as a sequence of $n-l+1$
$l$-mers, each overlapping the last by exactly $l-1$ bases. In this
definition we use members of the generalize decomposition, and find
they are not necessarily offset by just one base: given the $g_1$ and
$g_2$ example above, we see that the $r_1 = AAAAA*TTTTT$ descriptor is
composed of $AA*AA$, $AA*TT$, and $TT*TT$ -- each offset from the last
by 3.  As RAIDER assumed that it need a look-back of only one, this
too breaks the algorithm.

\subsection*{phRAIDER}

In order to adapt the RAIDER algorithm to the new definition, we
require observations parallel to those on which RAIDER was built.  As
it turns out, its easy to adopt those.  Consider, for example, the
first observation: ``An $l$-mer cannot belong to two different
elementary repeats.''  We have already seen a violation, but can
eliminate this by changing the definition of {\it belongs}:
\begin{definition}
  Given a elementary repeat descriptor $r$ for seed $s$ with length
  $l$, we say a length $w$ substring \underline{belongs} to $r$ if it
  matches a member of the generalized decomposition of $r$ (w.r.t. to
  $s$.)
\end{definition}
We notice in the above example, the string $AAAAA$ belongs to the
generalized decomposition, hence could not appear in both elementary
repeats.  But the string $AAAAT$ does not match any string in the
generalized decomposition of $r$, and hence was free to belong to
the descriptor of a ``distinct'' family.

\subsection*{phRAIDER Algorithm}



  
    









%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% Backmatter begins here                   %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\begin{backmatter}

\section*{Competing interests}
  The authors declare that they have no competing interests.

\section*{Author's contributions}
    Text for this section \ldots

\section*{Acknowledgements}
  Text for this section \ldots
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                  The Bibliography                       %%
%%                                                         %%
%%  Bmc_mathpys.bst  will be used to                       %%
%%  create a .BBL file for submission.                     %%
%%  After submission of the .TEX file,                     %%
%%  you will be prompted to submit your .BBL file.         %%
%%                                                         %%
%%                                                         %%
%%  Note that the displayed Bibliography will not          %%
%%  necessarily be rendered by Latex exactly as specified  %%
%%  in the online Instructions for Authors.                %%
%%                                                         %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% if your bibliography is in bibtex format, use those commands:
\bibliographystyle{bmc-mathphys} % Style BST file (bmc-mathphys, vancouver, spbasic).
\bibliography{bmc}      % Bibliography file (usually '*.bib' )
% for author-year bibliography (bmc-mathphys or spbasic)
% a) write to bib file (bmc-mathphys only)
% @settings{label, options="nameyear"}
% b) uncomment next line
%\nocite{label}

% or include bibliography directly:
% \begin{thebibliography}
% \bibitem{b1}
% \end{thebibliography}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                               %%
%% Figures                       %%
%%                               %%
%% NB: this is for captions and  %%
%% Titles. All graphics must be  %%
%% submitted separately and NOT  %%
%% included in the Tex document  %%
%%                               %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%
%% Do not use \listoffigures as most will included as separate files

%% \section*{Figures}
%%   \begin{figure}[h!]
%%   \caption{\csentence{Sample figure title.}
%%       A short description of the figure content
%%       should go here.}
%%       \end{figure}

%% \begin{figure}[h!]
%%   \caption{\csentence{Sample figure title.}
%%       Figure legend text.}
%%       \end{figure}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                               %%
%% Tables                        %%
%%                               %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%% Use of \listoftables is discouraged.
%%
%% \section*{Tables}
%% \begin{table}[h!]
%% \caption{Sample table title. This is where the description of the table should go.}
%%       \begin{tabular}{cccc}
%%         \hline
%%            & B1  &B2   & B3\\ \hline
%%         A1 & 0.1 & 0.2 & 0.3\\
%%         A2 & ... & ..  & .\\
%%         A3 & ..  & .   & .\\ \hline
%%       \end{tabular}
%% \end{table}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                               %%
%% Additional Files              %%
%%                               %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%% \section*{Additional Files}
%%   \subsection*{Additional file 1 --- Sample additional file title}
%%     Additional file descriptions text (including details of how to
%%     view the file, if it is in a non-standard format or the file extension).  This might
%%     refer to a multi-page table or a figure.

%%   \subsection*{Additional file 2 --- Sample additional file title}
%%     Additional file descriptions text.


\end{backmatter}
\end{document}

%%  LocalWords:  de addressref aff neteref Schaeffer noteref Xiaolin
%%  LocalWords:  Liu corref JE Karro Initio phRAIDER genomic TEs Ns
%%  LocalWords:  JIANG RepeatMasker nHMMER BLASTing homology ALU ReAS
%%  LocalWords:  LINEs Jiang PILER RepeatScout mer RepeatGluer Saha
%%  LocalWords:  infeasible DeBrujin et al Zheng Lonardi pre Huo mers
%%  LocalWords:  PatternHunter decompositions subsequence minimality
%%  LocalWords:  maximality RIADER ers substring AACACA AAGACA TACACA
%%  LocalWords:  TTAACTCA TTTTT TTT substrings AAATTT AAATT TT AAAAA
%%  LocalWords:  FIGUREOA AAAAAGTTTTT AAAAGTTTT AAAA TTTT AAAAT
%%  LocalWords:  Acknowledgements
