%% BioMed_Central_Tex_Template_v1.06
%%                                      %
%  bmc_article.tex            ver: 1.06 %
%                                       %

%%IMPORTANT: do not delete the first line of this template
%%It must be present to enable the BMC Submission system to
%%recognise this template!!

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                     %%
%%  LaTeX template for BioMed Central  %%
%%     journal article submissions     %%
%%                                     %%
%%          <8 June 2012>              %%
%%                                     %%
%%                                     %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                                                 %%
%% For instructions on how to fill out this Tex template           %%
%% document please refer to Readme.html and the instructions for   %%
%% authors page on the biomed central website                      %%
%% http://www.biomedcentral.com/info/authors/                      %%
%%                                                                 %%
%% Please do not use \input{...} to include other tex files.       %%
%% Submit your LaTeX manuscript as one .tex document.              %%
%%                                                                 %%
%% All additional figures and files should be attached             %%
%% separately and not embedded in the \TeX\ document itself.       %%
%%                                                                 %%
%% BioMed Central currently use the MikTex distribution of         %%
%% TeX for Windows) of TeX and LaTeX.  This is available from      %%
%% http://www.miktex.org                                           %%
%%                                                                 %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%% additional documentclass options:
%  [doublespacing]
%  [linenumbers]   - put the line numbers on margins

%%% loading packages, author definitions

%\documentclass[twocolumn]{bmcart}% uncomment this for twocolumn layout and comment line below
\documentclass{bmcart}

%%% Load packages
%\usepackage{amsthm,amsmath}
%\RequirePackage{natbib}
%\RequirePackage[authoryear]{natbib}% uncomment this for author-year bibliography
%\RequirePackage{hyperref}
\usepackage[utf8]{inputenc} %unicode support
%\usepackage[applemac]{inputenc} %applemac support if unicode package fails
%\usepackage[latin1]{inputenc} %UNIX support if unicode package fails


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                             %%
%%  If you wish to display your graphics for   %%
%%  your own use using includegraphic or       %%
%%  includegraphics, then comment out the      %%
%%  following two lines of code.               %%
%%  NB: These line *must* be included when     %%
%%  submitting to BMC.                         %%
%%  All figure files must be submitted as      %%
%%  separate graphics through the BMC          %%
%%  submission process, not included in the    %%
%%  submitted article.                         %%
%%                                             %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


\def\includegraphic{}
\def\includegraphics{}



%%% Put your definitions there:
\startlocaldefs
\endlocaldefs


%%% Begin ...
\begin{document}

%%% Start of article front matter
\begin{frontmatter}

\begin{fmbox}
\dochead{Research}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% Enter the title of your article here     %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\title{Fast de novo transpoible element annotation}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% Enter the authors here                   %%
%%                                          %%
%% Specify information, if available,       %%
%% in the form:                             %%
%%   <key>={<id1>,<id2>}                    %%
%%   <key>=                                 %%
%% Comment or delete the keys which are     %%
%% not used. Repeat \author command as much %%
%% as required.                             %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\author[
   addressref={aff1},                   % id's of addresses, e.g. {aff1,aff2}
   noteref={n1},                        % id's of article notes, if any
   email={figuernd@miamiOH.edu}   % email address
]{\inits{ND}\fnm{Nathaniel D} \snm{Figueroa}}
\author[
   addressref={aff1},
   email={schaefce@miamiOH.edu}
]{\inits{SE}\fnm{Carly E} \snm{Schaeffer}}
\author[
   addressref={aff1,aff2},
   email={liux17@miamiOH.edu}
]{\inits{X}\fnm{Xiaolin} \snm{Liu}}
\author[
  addressref={aff1,aff2,aff3,aff4},
  corref={aff1},
  noteref={n2},
  email={karroje@miamiOH.edu}
]{\inits{JE}\fnm{John E} \snm{Karro}}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% Enter the authors' addresses here        %%
%%                                          %%
%% Repeat \address commands as much as      %%
%% required.                                %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\address[id=aff1]{%                           % unique id
  \orgname{Department of Computer Science and Software Engineering}, % university, etc
  %\street{Miami University},                     %
  %\postcode{}                                % post or zip code
  %\city{Oxford},                              % city
  %\cny{USA}                                    % country
}
\address[id=aff2]{%                           % unique id
  \orgname{Center for  Molecular and Structural Biology},
  %\street{Miami University},                     %
  %\postcode{}                                % post or zip code
  %\city{Oxford},                              % city
  %\cny{USA}                                    % country
}
\address[id=aff3]{%                           % unique id
  \orgname{Department of Statistics}, % university, etc
  %\street{Miami University},                     %
  %\postcode{}                                % post or zip code
  %\city{Oxford},                              % city
  %\cny{USA}                                    % country
}
\address[id=aff4]{%                           % unique id
  \orgname{Department of Microbiology}, % university, etc
  \street{Miami University},                     %
  %\postcode{}                                % post or zip code
  \city{Oxford},                              % city
  \cny{USA}                                    % country
}


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% Enter short notes here                   %%
%%                                          %%
%% Short notes will be after addresses      %%
%% on first page.                           %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\begin{artnotes}
%\note{Sample of title note}     % note to the article
  \note[id=n1]{Equal contributor} % note, connected to author
  \note[id=n2]{Corresponding Author}
\end{artnotes}

\end{fmbox}% comment this for two column layout

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% The Abstract begins here                 %%
%%                                          %%
%% Please refer to the Instructions for     %%
%% authors on http://www.biomedcentral.com  %%
%% and include the section headings         %%
%% accordingly for your article type.       %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\begin{abstractbox}

\begin{abstract} % abstract
\parttitle{First part title} %if any
Text for this section.

\parttitle{Second part title} %if any
Text for this section.
\end{abstract}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% The keywords begin here                  %%
%%                                          %%
%% Put each keyword in separate \kwd{}.     %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\begin{keyword}
\kwd{sample}
\kwd{article}
\kwd{author}
\end{keyword}

% MSC classifications codes, if any
%\begin{keyword}[class=AMS]
%\kwd[Primary ]{}
%\kwd{}
%\kwd[; secondary ]{}
%\end{keyword}

\end{abstractbox}
%
%\end{fmbox}% uncomment this for twcolumn layout

\end{frontmatter}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% The Main Body begins here                %%
%%                                          %%
%% Please refer to the instructions for     %%
%% authors on:                              %%
%% http://www.biomedcentral.com/info/authors%%
%% and include the section headings         %%
%% accordingly for your article type.       %%
%%                                          %%
%% See the Results and Discussion section   %%
%% for details on how to create sub-sections%%
%%                                          %%
%% use \cite{...} to cite references        %%
%%  \cite{koon} and                         %%
%%  \cite{oreg,khar,zvai,xjon,schn,pond}    %%
%%  \nocite{smith,marg,hunn,advi,koha,mouse}%%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\newtheorem{definition}{Definition}
\newtheorem{observation}{Observation}
\newtheorem{lemma}{Lemma}

%%%%%%%%%%%%%%%%%%%%%%%%% start of article main body
\section*{Background}
Here we introduce phRAIDER (Rapid Ab Initio Detection of Elementary
Repeat), designed to enable the quick annoation of transposible
elements and repetitive DNA without requiring a pre-compiled template
library.  Transposible elements, or TEs, are present in almost every
higher order genome, covering as much as 45\% of the human genome and
90\% of the maize genome [CHECK THIS] [CITE].  Given their prevelence,
identiication of TEs is important not only for those who study them, but for
anyone trying to understand a geome.  If not filtered, TEs can trigger
hug numbers of false positives in the course of applying automated
gene finding techniques, as well as blow up runtime.  [CITE: Jiang]
So identifying and masking them in a must for any genome being
studied.

The most well known tools for repeat identification ar RepeatMasker
and [Eddy's tool] [CITE], both of which work on the principle of a
library-based search.  That is, to find transposable elements of a
given family, the tools require some description of sequence in the
family (in the form of a model sequence for blasting, or a pHMM) that
is used as the basis for identification.  This begs the obvious
question: much like we ask how the snow plow driver gets to work, or
how dictinary authors look up the spelling of a word [CITE], we must ask how the
tools compile their libraries.  How do we compile the libraries of
transposible elements, or discover new transpoable element families.

Within mammalian speices we can, to some extent, rely on homology
relationships to port libraries across species.  A library description
of a human ALU sequence can likely be used for the initial annotation
of a newly sequence primate sequence, and many LINEs will extend to
rodent, or even more distantn mammalian species.  This does not hold
so well in plants.  In many cases TE composition of a given organism
is species-specific.  For example, a rice-based TE libarary will only identify 25\%
of the TEs in the maize genome. [CITE Jiang] 

To solve this problem we turn to {\it de novo} TE identification
tools: tools identifying TEs though a genome using only the genome
sequence information.  A number of such tls have been publised in the
literature, operating on a number of different principles.  Tools such
as RECON and PILER are based on self-alignment, using WU-BLAST and
PALS for the alignemnt [CITE].  RECON show good sensitivity but is
computationally intesive and infeasible for use on whole genomes (requiring 60 hours for 18Mb rice genome in a 2013
report), while PILER achieves a good runtime with very low
sensitivey. [CITE JIANG].   ReAS and RepeatScout are based on $k$-mer
searches [CITE], with the earlier showing less sensitivity than
RECON.  Finally RepeatGluer is based on a vairation od DeBrujin
grauphs [CITE], which allows for a decomposition of TE families into
domains, but is very, very computationally expensive.  In Saha {\it et
  al.} the authors perform an extenive comparison of the tools, and
conclude that RepeatScout is the best tool overall for assembled
genomes and ReAS the best when dealing with unassembled sequeunce
fragments [CITE].

\subsection*{Elementary Repeats}
Another line of developement, first proposed by Zheng and Lonardi,
involves the use of {\it elementary repeats} [CITE].  
Zheng and Lonardi formulated their defintion with the explicit intention of factoring minimum length and
frequency into the definition: to be an elementary repeat a sequence
must have some minimum length $l$ and occur at least $f$ times.  More
importantly, it must be minimal (that is, no section of an elementary
repeat can itself be an elementary repeat), and it must have appear
independently of any other sequence meeting definition of an
elementary repeat.  [CITE]  Formally:
\begin{definition}
Give a length criteria $l$ and a frequence criteria $f$, a sequence
$r$ is an elementary repeat in a genome if:
\begin{enumerate}
\label{ZLDef}
\item $r$ is of length at least $l$.
\item There are at least $f$ copies of $r$ in the genome.
\item There is no proper subsequence $r'$ of $r$ such that $r'$ is of
  length $l$ or grater and $r'$ occurs independently of $r$.  (That
  is, every length $>l$ subsequence of $r'$ occurs only where $r'$
  occurs in the genome.  (The minimality condition.)
\item There is no sequence $r'$ that properly contains $r$ and appears
  around every occurence of $r$ in the genome.
\end{enumerate}
\end{definition}
The third condition insures minimality (that an elementary repeat
cannot contain elementary repeats), while the fourth condition that
each elementary repeat is maximal in length.  Having proposed this
definition, Zheng and Lonardi developed an identification algorithm
that had a runtime quadratic in the genome size -- so not of practical
use on whole-genome analysis.  This was refined to linear time by He [CITE],
and also by Huo {\it et al.} [CITE] based on variations suffix tree
approaches, but because of that use of suffix trees they are limited
in their ability to handle sequence variation.  As transposible
elements naturallly accumulate instance-specific base substitutions
over time, this is a significant limitation.

\subsection*{RAIDER}
It was with the objective of creating a linear time identification
algorithm that could handle variation through use of
PatternHunter-like spaced seeds [CITE].  A rough implementatin was
first presentd in Figuera {\it et al.}, with more details in the
Figuera masters thesis [CITE].  Not itself fully able to handle the
spaced seed component, RAIDER was built along an alternate,
but equivilent definition of elementary repeats based on $l$-mers
(sequences of length $l$).  Specifically, it was observed that
condition (3) of Definition~\ref{ZLDef} could be rewritten as: There
is no $l$-mer contained with $r$ that appears in the genome
independelty of $r$.

From this it follows that:
\begin{enumerate}
\item An $l$-mer cannot belong to two different elementary repeats.
\item Any $l$-mer in the genome that occurs $f$ or more times is
  either an elementary repeat or is contained on one.
\item Any two $l$-ers contained within an elementary repeat must occur
  the same number of times in the genome.
\item Given two sequences in the genome that are {\it maximally
  idenitcal} (that is, cannot be extended in either direction and stil
  be identical), these sequences cannot be contained with in
  elementart repeat.
\end{enumerate}
(For dicussion and proof, see the Figueroa Thesis [CITE]].)

Based on these observations, we discover we can find all elementary
repeats in a single scan of the genome.  Specifically, as we scan from
left to right, we track $l$-mer occurances, and look for repeated
$l$-mer sequences.  When we find the same sequence of $l$-mers
occuring multile times in a row, we can mark it as a tentative family,
then brack it down later if we discover violations of the minimality
condition.  The alogrithm is summerized in Figueroa {\it et al.} [CITE], and
discussed in detail in the Figueroa Thesis [CITE].

Results of the preliminary implementation were promising.  Based on
the Saha {\it et al} analysis putting RepeatMasker as the top {\it de
  novo} search tool, comparisone were aginst that.  RAIDER result
quality was close, but slightly behind -- with runtime
order of magnitudes better.  On human chromosome 22 we saw a
$12\times$ speedup (2344 seconds to 192 seconds), while coverage
actually improved (77\% to 84\%), while on mouse chromsome 19 we saw 
the same speedup with a sinifcant drop in coverage (53\% to 30\%).  On
the full human genome RAIDER ran in $6.3$ hours, while RepeatScout was
unable to complete it run.  For details, see Figueroa {\it et al.} [CITE]

\subsection*{Spaced Seeds}
 
PatternHunter, first proposed as a very succesful augmentation to
BLAST [CITE PatternHunter,BLAST], is based on the notion of {\it
  spaced seeds}.  In the context of BLAST, instead of looking for a
cntiguous substring match to triger the alignment exention,
PatternHunter uses a more refised match pattern.  For example, the
{\it seed} 1111110111111 would specify that we were looking for an exact
match of two length-six substrings seperated by a single base (which
might or might not match).  In the toy seed $s=110101$, we would say the
strings $AACACA$ and $AAGACA$ because all positions corresponding to a
1 match.  The string $AACACA$ does not math $TACACA$ (with respect to
$s$), as the single miss-match does not correspond to the 1.  Nor does
$AACACA$ match ant substring of $TTAACTCA$.

The incorperations of spaced seeds into BLAST lead to significant
improvements at vertual no cost in runtime [CITE], so it made sence to
incorportat them here.  Specifcially, instead of looking at repeat
elements as a set of identical strings, we could require that they
match with respect to a PatternHunter spaced seed.  The preliminary
version of RAIDER from Figueroa {\it et al.} made some attempt to
incorperate this -- was succesful enough to provide a
proof-of-concept, but a significant amount of work was left to be done
for the developement of phRAIDER (PatternHunter-based RAIDER) -- the
main method in this paper.

\section{Methods}









%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% Backmatter begins here                   %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\begin{backmatter}

\section*{Competing interests}
  The authors declare that they have no competing interests.

\section*{Author's contributions}
    Text for this section \ldots

\section*{Acknowledgements}
  Text for this section \ldots
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                  The Bibliography                       %%
%%                                                         %%
%%  Bmc_mathpys.bst  will be used to                       %%
%%  create a .BBL file for submission.                     %%
%%  After submission of the .TEX file,                     %%
%%  you will be prompted to submit your .BBL file.         %%
%%                                                         %%
%%                                                         %%
%%  Note that the displayed Bibliography will not          %%
%%  necessarily be rendered by Latex exactly as specified  %%
%%  in the online Instructions for Authors.                %%
%%                                                         %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% if your bibliography is in bibtex format, use those commands:
\bibliographystyle{bmc-mathphys} % Style BST file (bmc-mathphys, vancouver, spbasic).
\bibliography{bmc_article}      % Bibliography file (usually '*.bib' )
% for author-year bibliography (bmc-mathphys or spbasic)
% a) write to bib file (bmc-mathphys only)
% @settings{label, options="nameyear"}
% b) uncomment next line
%\nocite{label}

% or include bibliography directly:
% \begin{thebibliography}
% \bibitem{b1}
% \end{thebibliography}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                               %%
%% Figures                       %%
%%                               %%
%% NB: this is for captions and  %%
%% Titles. All graphics must be  %%
%% submitted separately and NOT  %%
%% included in the Tex document  %%
%%                               %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%
%% Do not use \listoffigures as most will included as separate files

\section*{Figures}
  \begin{figure}[h!]
  \caption{\csentence{Sample figure title.}
      A short description of the figure content
      should go here.}
      \end{figure}

\begin{figure}[h!]
  \caption{\csentence{Sample figure title.}
      Figure legend text.}
      \end{figure}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                               %%
%% Tables                        %%
%%                               %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%% Use of \listoftables is discouraged.
%%
\section*{Tables}
\begin{table}[h!]
\caption{Sample table title. This is where the description of the table should go.}
      \begin{tabular}{cccc}
        \hline
           & B1  &B2   & B3\\ \hline
        A1 & 0.1 & 0.2 & 0.3\\
        A2 & ... & ..  & .\\
        A3 & ..  & .   & .\\ \hline
      \end{tabular}
\end{table}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                               %%
%% Additional Files              %%
%%                               %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\section*{Additional Files}
  \subsection*{Additional file 1 --- Sample additional file title}
    Additional file descriptions text (including details of how to
    view the file, if it is in a non-standard format or the file extension).  This might
    refer to a multi-page table or a figure.

  \subsection*{Additional file 2 --- Sample additional file title}
    Additional file descriptions text.


\end{backmatter}
\end{document}
